File: data/derived_cot/rq1_traces/translation/qwen/java_to_cpp/code_output/AccessGatewayFilter.cpp
#include <unordered_map>
#include <any>
#include <string>
#include <vector>
#include <stdexcept>
#include <ctime>
#include <iostream>
#include <iomanip>
#include <sstream>

// Helper function to safely retrieve and cast values from a map
template <typename T>
T getFromMap(const std::unordered_map<std::string, std::any>& map, const std::string& key) {
    auto it = map.find(key);
    if (it == map.end()) {
        throw std::runtime_error("Key not found: " + key);
    }
    try {
        return std::any_cast<T>(it->second);
    } catch (const std::bad_any_cast& e) {
        throw std::runtime_error("Invalid type for key '" + key + "': expected type of " + 
                               typeid(T).name());
    }
}

class AccessGatewayFilter {
public:
    AccessGatewayFilter() {}

    bool filter(std::unordered_map<std::string, std::any> request) {
        try {
            std::string requestUri = getFromMap<std::string>(request, "path");
            std::string method = getFromMap<std::string>(request, "method");

            if (isStartWith(requestUri)) {
                return true;
            }

            try {
                std::unordered_map<std::string, std::any> token = getJwtUser(request);
                std::unordered_map<std::string, std::any> user = getFromMap<std::unordered_map<std::string, std::any>>(token, "user");
                int level = getFromMap<int>(user, "level");
                if (level > 2) {
                    setCurrentUserInfoAndLog(user);
                    return true;
                }
            } catch (...) {
                return false;
            }
            return false;
        } catch (...) {
            return false;
        }
    }

    bool isStartWith(const std::string& requestUri) {
        std::vector<std::string> startWith = {"/api", "/login"};
        for (const auto& s : startWith) {
            if (requestUri.substr(0, s.length()) == s) {
                return true;
            }
        }
        return false;
    }

    std::unordered_map<std::string, std::any> getJwtUser(const std::unordered_map<std::string, std::any>& request) {
        try {
            std::unordered_map<std::string, std::any> headers = getFromMap<std::unordered_map<std::string, std::any>>(request, "headers");
            std::unordered_map<std::string, std::any> tokenMap = getFromMap<std::unordered_map<std::string, std::any>>(headers, "Authorization");
            std::unordered_map<std::string, std::any> user = getFromMap<std::unordered_map<std::string, std::any>>(tokenMap, "user");
            std::string jwt = getFromMap<std::string>(tokenMap, "jwt");
            std::string name = getFromMap<std::string>(user, "name");

            if (jwt.substr(0, name.length()) == name) {
                std::string jwtDateStr = jwt.substr(name.length());
                std::tm t = {};
                char* ptr = std::strptime(jwtDateStr.c_str(), "%Y-%m-%d", &t);
                if (ptr == nullptr || ptr != jwtDateStr.c_str() + jwtDateStr.size()) {
                    return {};
                }

                std::time_t time_val = std::mktime(&t);
                std::time_t now = std::time(nullptr);
                std::tm today = *std::localtime(&now);
                today.tm_mday -= 3;
                std::mktime(&today); // Normalize
                std::time_t three_days_ago = std::mktime(&today);

                if (time_val < three_days_ago) {
                    return {};
                }
            }
            return tokenMap;
        } catch (...) {
            return {};
        }
    }

    void setCurrentUserInfoAndLog(const std::unordered_map<std::string, std::any>& user) {
        std::string host = getFromMap<std::string>(user, "address");
        std::string name = getFromMap<std::string>(user, "name");
        std::string currentDate = getCurrentDate();
        std::cout << name << host << currentDate << std::endl;
    }

    std::string getCurrentDate() {
        std::time_t now = std::time(nullptr);
        std::tm* now_tm = std::localtime(&now);
        char buffer[11];
        std::strftime(buffer, sizeof(buffer), "%Y-%m-%d", now_tm);
        return std::string(buffer);
    }
};

File: data/derived_cot/rq1_traces/translation/r1/java_to_py/code_output/AccessGatewayFilter.py
from datetime import datetime, timedelta, date

class AccessGatewayFilter:
    def __init__(self):
        pass

    def filter(self, request):
        request_uri = request.get("path")
        method = request.get("method")

        if self.isStartWith(request_uri):
            return True

        try:
            token = self.getJwtUser(request)
            if token is None:
                return False
            user = token.get("user")
            if user is None:
                return False
            if user.get("level") > 2:
                self.setCurrentUserInfoAndLog(user)
                return True
        except Exception:
            return False
        return False

    def isStartWith(self, request_uri):
        start_with = ["/api", "/login"]
        for s in start_with:
            if request_uri.startswith(s):
                return True
        return False

    def getJwtUser(self, request):
        headers = request.get("headers")
        if headers is None:
            return None
        token = headers.get("Authorization")
        if token is None:
            return None
        user = token.get("user")
        jwt_str = token.get("jwt")
        if user is None or jwt_str is None:
            return token

        user_name = user.get("name")
        if user_name is None:
            return token

        if jwt_str.startswith(user_name):
            jwt_date_str = jwt_str[len(user_name):]
            try:
                jwt_date = datetime.strptime(jwt_date_str, "%Y-%m-%d").date()
            except ValueError:
                return token
            if date.today() - timedelta(days=3) > jwt_date:
                return None
        return token

    def setCurrentUserInfoAndLog(self, user):
        host = user.get("address", "")
        user_name = user.get("name", "")
        today = date.today().strftime("%Y-%m-%d")
        message = f"{user_name}{host}{today}"
        print(message)